src/digraph/Dominator_Tree.jl

```diff
@@ -0,0 +1,330 @@
+function dominator_Tree(g::AG, source) where {T, AG<:DiGraph{T}}
+
+
+
+  parent=zeros(T,nv(g)) #the parents of every node in the dfs teree
+
+
+  #= the semi dominator is the first node on the path between the root and the node of interest,
+  such that  there is at least two disjoint paths between it and the node of interest ,
+  if there is not such a node we will consider the parent of the node of interst to be the semi dominator of it
+   there is a duality between the semi dominator and the  immediate dominator
+   min(preorderof(semi_dominator(v)))>=max(preorder(dominator(v))
+   for more details see the paper
+   the key theorm is
+   Let w != the source, and let u be a vertex for which semi_dominator(u) is minimum
+   among vertices between w and its semi dominator.
+   Then the immediate dominator is
+   semi_dominator(w) if semi_dominator(w) = semi_dominator(u),
+   immidiate_dominator(w) =  immidiate_dominator(u) otherwise.
+  =#
+  semi=zeros(T,nv(g))
+  #==#
+
+
+  function accumfunc(a1,a2)
+    if semi[a1]<=semi[a2]
+      return true
+    else
+      return false
+    end
+  end
+      eval,link=produce_eval_link(nv(g),accumfunc,T)
+
+  #=
+   we will store each node in the bucket of its semi dominator,
+   until we we figure out all the semi dominator of the nodes between them
+  =#
+
+  bucktes=Array{Vector,1}(undef,nv(g))
+
+  verticesar=zeros(T,nv(g)) #map between each preorder of a node and its real name
+  dom=zeros(T,nv(g))        #the immediate dominator of the the node
+  #=
+  one important step in the algorithm is to know a  vertex u for which semi_dominator(u) is minimum
+  among vertices between every node and its semi dominator,
+  after we order the vertices by preorder, we will scan them backwards,
+  then we can imagine the graph as a forest of trees, in each step we merge two trees by making the root of one tree the
+  child of the root of the other tree, now we discovered  another child of the later root,let it be w
+  then we will see if there is node u that consider w to be its semi dominator, and look for the smallest smei dominator in the path between w and u
+  we can leverage some facts
+  - we donot need to travel all the path each time, we can do path decompsion
+  - each time we do a merge one of the two trees will not be affected at all,
+  in the other tree if there is a node that has semi dominator smaller than the new root its descendants will not be affected
+
+  =#
+
+
+
+  parent[source]=1
+
+  counter=Int(0)
+
+
+  
+    s=Vector{}([])
+    ptr=1
+    counter=1
+    semi[source]=1
+    verticesar[1]=source
+    bucktes[source]=Vector{}([])
+    push!(s,(source,0))
+    while !isempty(s)
+        v,_=s[end]
+      
+        vneighbors=outneighbors(g,v)
```

Comment:
This looks like a standard BFS. Is it? If so, please don't reinvent the wheel: we're just about to promote `traverse_graph!()` from `LightGraphs.Experimental.Traversals` which gives you BFS with state modifiers. Let's talk about what this algorithm does in the comments before we do any more coding.

Reply:
It is a dfs, ,the goal of this code is to map each node to dfs preorder number, and to meap every number in [1 ......  #nodes reachable from the source] to the node that has that number as its preorder number, 
I didn’t find any function that does that in the documentation, so i decided to write my own, i didn’t know about the experimental version of the dfs , because it wasn’t in documentation,i think if i am allowed to pass previsitfn as an argument to traverse_graph it will do the trick for me.